### DISCUSSION TITLE
how to add new method to Color

### DISCUSSION THREAD

**Discussion by Novakasa** [pybricks/support#1305] [Q&A] [answered]
Hey all, I am currently trying to expose the new hsv distance function to user programs (potentially for a pull request), so I looked at different implementations in pybricks for how methods are added to python classes in C. It looked to me like I had to use various macros to define the function object, define a `locals` dictionary that contains that object and then add that to the class definition.
This compiles, however trying to acces the method in a user program it seems like the `Color` instance still has no attribute with the name of the function. Is it maybe because of the custom attribute handler of `Color`?
You can see my implementation here:
https://github.com/pybricks/pybricks-micropython/compare/master...Novakasa:pybricks-micropython:expose-colordist
In the newest commit, I renamed the funcion to "dist" just to see if there's an issue with the underscore or something. What am I missing to add a new method to the `Color` class? Also I am having a hard time finding documentation on how to do this in micropython in the official micropython docs, is there a good ressource on this somewhere?

**Comment by laurensvalk** [comment, marked as answer]
The `Color` class has its own attribute handler `pb_type_Color_attr`.

By default, MicroPython will not look in the `locals_dict` if the `attr` function exists.

Here's the relevant MicroPython code:
```c
        // If type->attr has set dest[1] = MP_OBJ_SENTINEL, we should proceed
        // with lookups below (i.e. in locals_dict). If not, return right away.
        if (dest[1] != MP_OBJ_SENTINEL) {
            return;
        }
```

So you can update `pb_type_Color_attr` to set `dest[1] = MP_OBJ_SENTINEL` if the attribute lookup is not successful. Then it should proceed to your new `locals_dict`.

This is a feature we've upstreamed ourselves because we were frequently using both an attribute handler and a `locals_dict`, much like you are trying to do now :smile:

**Reply by Novakasa** [reply]
Thanks, that looks like the solution, will give it a shot.
